### PR TITLE
serviceData and manufacturerData support

### PR DESCRIPTION
**Pull request by zakorgy**
Allow requesting for BluetoothDevices with service and manufacturer specific data.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14277)

### CODE REVIEW

File: components/script/dom/bluetooth.rs [hunk 1]
```diff
@@ -285,31 +292,72 @@ fn canonicalize_filter(filter: &BluetoothRequestDeviceFilter) -> Fallible<Blueto
...
-                                manufacturer_id,
-                                service_data_uuid))
+    // Step 10.
+    Ok(BluetoothScanfilter::new(name, name_prefix, services_vec, manufacturer_data, service_data))
+}
+
+// https://webbluetoothcg.github.io/web-bluetooth/#bluetoothdatafilterinit-canonicalizing
+fn canonicalize_bluetooth_data_filter_init(bdfi: &BluetoothDataFilterInit) -> Fallible<(Vec<u8>, Vec<u8>)> {
+    // Step 1.
+    let data_prefix = match bdfi.dataPrefix {
```
Comment (jdm) [inline components/script/dom/bluetooth.rs]: `let data_prefix = bdfi.dataPrefix.cloned().unwrap_or(vec![]);`

File: components/script/dom/bluetooth.rs [hunk 2]
```diff
@@ -285,31 +292,72 @@ fn canonicalize_filter(filter: &BluetoothRequestDeviceFilter) -> Fallible<Blueto
...
-            if uuid_is_blacklisted(uuid.as_ref(), Blacklist::All) {
-                return Err(Security)
+    // Step 2.4.6 - 2.4.7
+    let manufacturer_data = match filter.manufacturerData {
+        Some(ref manufacturer_data_map) => {
+            if manufacturer_data_map.deref().is_empty() {
+                return Err(Type(MANUFACTURER_DATA_ERROR.to_owned()));
             }
+            let mut map = HashMap::new();
+            for (key, bdfi) in manufacturer_data_map.deref() {
```
Comment (jdm) [inline components/script/dom/bluetooth.rs]: Can't this be `for (key, bdfi) in &manufacturer_data_map {`?

File: components/script/dom/bluetooth.rs [hunk 3]
```diff
@@ -285,31 +292,72 @@ fn canonicalize_filter(filter: &BluetoothRequestDeviceFilter) -> Fallible<Blueto
...
+        },
+        None => None,
+    };
 
-            // Step 2.4.7.4.
-            uuid
+    // Step 2.4.8 -2.4.9
+    let service_data = match filter.serviceData {
+        Some(ref service_data_map) => {
+            if service_data_map.deref().is_empty() {
```
Comment (jdm) [inline components/script/dom/bluetooth.rs]: No need for deref.

File: components/script/dom/bluetooth.rs [hunk 4]
```diff
@@ -285,31 +292,72 @@ fn canonicalize_filter(filter: &BluetoothRequestDeviceFilter) -> Fallible<Blueto
...
+                return Err(Type(SERVICE_DATA_ERROR.to_owned()));
+            }
+            let mut map = HashMap::new();
+            for (key, bdfi) in service_data_map.deref() {
```
Comment (jdm) [inline components/script/dom/bluetooth.rs]: `for (key, bdfi) in &service_data_map {`

File: components/bluetooth_traits/scanfilter.rs [hunk 5]
```diff
@@ -60,20 +63,20 @@ impl BluetoothScanfilter {
         &self.services.0
     }
 
-    pub fn get_manufacturer_id(&self) -> Option<u16> {
-        self.manufacturer_id
+    pub fn get_manufacturer_data(&self) -> &Option<ManufacturerData> {
```
Comment (jdm) [inline components/bluetooth_traits/scanfilter.rs]: `-> Option<&ManufacturerData>`

File: components/bluetooth_traits/scanfilter.rs [hunk 6]
```diff
@@ -60,20 +63,20 @@ impl BluetoothScanfilter {
...
+        &self.manufacturer_data
```
Comment (jdm) [inline components/bluetooth_traits/scanfilter.rs]: `self.manufacturer_data.as_ref()`

File: components/bluetooth_traits/scanfilter.rs [hunk 7]
```diff
@@ -60,20 +63,20 @@ impl BluetoothScanfilter {
...
     }
 
-    pub fn get_service_data_uuid(&self) -> &str {
-        &self.service_data_uuid
+    pub fn get_service_data(&self) -> &Option<ServiceData> {
+        &self.service_data
```
Comment (jdm) [inline components/bluetooth_traits/scanfilter.rs]: `self.service_data.as_ref()`, and change the return type.

File: components/bluetooth/lib.rs [hunk 8]
```diff
@@ -130,26 +130,63 @@ fn matches_filter(device: &BluetoothDevice, filter: &BluetoothScanfilter) -> boo
...
+// https://webbluetoothcg.github.io/web-bluetooth/#bluetoothdatafilterinit-matches
+fn data_filter_matches(data: &[u8], prefix: &[u8], mask: &[u8]) -> bool {
+    // Step 1-2: No need to copy the bytes here.
+    // Step 3.
+    if data.len() < prefix.len() {
+        return false;
+    }
+
+    // Step 4.
+    for i in 0..mask.len() {
```
Comment (jdm) [inline components/bluetooth/lib.rs]:
```rust
for ((data, mask), prefix) in data.iter().zip(mask.iter()).zip(prefix.iter()) {
    if data & mask != prefix & mask {
        return false;
    }
}
```

File: tests/wpt/mozilla/tests/mozilla/bluetooth/requestDevice/canonicalizeFilter/wrong-service-data-key.html [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+}];
+
+promise_test(t => {
+    window.testRunner.setBluetoothMockDataSet(adapter_type.glucose_heart_rate);
+    let promises = [];
+    wrong_service_data_keys.forEach(args => {
+        promises.push(promise_rejects(t, 'SyntaxError', window.navigator.bluetooth.requestDevice(args)));
+    });
+    return Promise.all(promises);
+}, 'Rejects with SyntaxError, if a serviceData key not converts to a valid uuid.');
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/requestDevice/canonicalizeFilter/wrong-service-data-key.html]: `if a serviceData key does not convert to a valid uuid`.

File: tests/wpt/mozilla/tests/mozilla/bluetooth/requestDevice/device-not-found-with-service-and-manufacturer-data.html [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        window.navigator.bluetooth.requestDevice({
+            filters: [{
+                manufacturerData: {
+                    17: {dataPrefix: new Uint8Array([1, 2, 3])},
+                },
+                serviceData: {
+                    'heart_rate': {dataPrefix: new Uint8Array([1, 2, 3])}
+                }
+            }]}));
+}, 'Requesting with manufacturerData and serviceData in the same filter, will not find a device which does not advertises both.');
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/requestDevice/device-not-found-with-service-and-manufacturer-data.html]: `does not advertise`

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/bluetooth.rs (hunk 1 above)
Comment (dati91) [inline components/script/dom/bluetooth.rs]: Note: I changed this to `clone` not `cloned`. I'm not sure why was the `ref` used here.

**Comment by jdm** [comment]
Looks good! Go ahead and squash the commits, then this can be merged.
@bors-servo: delegate+

**Comment by KiChjang** [comment]
@bors-servo retry
* #14338

**Pull request merged**